[SPARK-17658][SPARKR] read.df/write.df API taking path optionally in SparkR

HyukjinKwon opened the pull request:
## What changes were proposed in this pull request?

`write.df`/`read.df` API require path which is not actually always necessary in Spark. Currently, it only affects the datasources implementing `CreatableRelationProvider`. Currently, Spark currently does not have internal data sources implementing this but it'd affect other external datasources.

In addition we'd be able to use this way in Spark's JDBC datasource after https://github.com/apache/spark/pull/12601 is merged.

**Before**
- `read.df`
  
  ``` r
  > read.df(source = "json")
  Error in dispatchFunc("read.df(path = NULL, source = NULL, schema = NULL, ...)",  :
  argument "x" is missing with no default
  ```
  
  ``` r
  > read.df(path = c(1, 2))
  Error in dispatchFunc("read.df(path = NULL, source = NULL, schema = NULL, ...)",  :
  argument "x" is missing with no default
  ```
  
  ``` r
  > read.df(c(1, 2))
  Error in invokeJava(isStatic = TRUE, className, methodName, ...) :
  java.lang.ClassCastException: java.lang.Double cannot be cast to java.lang.String
    at org.apache.spark.sql.execution.datasources.DataSource.hasMetadata(DataSource.scala:300)
    at
  ...
  In if (is.na(object)) { :
  ...
  ```
- `write.df`
  
  ``` r
  > write.df(df, source = "json")
  Error in (function (classes, fdef, mtable)  :
  unable to find an inherited method for function ‘write.df’ for signature ‘"function", "missing"’
  ```
  
  ``` r
  > write.df(df, source = c(1, 2))
  Error in (function (classes, fdef, mtable)  :
  unable to find an inherited method for function ‘write.df’ for signature ‘"SparkDataFrame", "missing"’
  ```
  
  ``` r
  > write.df(df, mode = TRUE)
  Error in (function (classes, fdef, mtable)  :
  unable to find an inherited method for function ‘write.df’ for signature ‘"SparkDataFrame", "missing"’
  ```

**After**
- `read.df`
  
  ``` r
  > read.df(source = "json")
  Error in loadDF : analysis error - Unable to infer schema for JSON at . It must be specified manually;
  ```
  
  ``` r
  > read.df(path = c(1, 2))
  Error in f(x, ...) : path should be charactor, null or omitted.
  ```
  
  ``` r
  > read.df(c(1, 2))
  Error in f(x, ...) : path should be charactor, null or omitted.
  ```
- `write.df`
  
  ``` r
  > write.df(df, source = "json")
  Error in save : illegal argument - 'path' is not specified
  ```
  
  ``` r
  > write.df(df, source = c(1, 2))
  Error in .local(df, path, ...) :
  source should be charactor, null or omitted. It is 'parquet' by default.
  ```
  
  ``` r
  > write.df(df, mode = TRUE)
  Error in .local(df, path, ...) :
  mode should be charactor or omitted. It is 'error' by default.
  ```
## How was this patch tested?

Unit tests in `test_sparkSQL.R`

HyukjinKwon committed:
Make path optional in write.df

HyukjinKwon commented:
cc @felixcheung and @shivaram

SparkQA commented:
**[Test build #65865 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65865/consoleFull)** for PR 15231 at commit [`41611e5`](https://github.com/apache/spark/commit/41611e594c4e5f8e6aedb2846abfda74dac84a80).

HyukjinKwon commented:
FWIW - Python API seems having `path` as optional[1] and Scala API seems doing this so as well[2].

[1]https://github.com/apache/spark/blob/master/python/pyspark/sql/readwriter.py#L521
[2]https://github.com/apache/spark/blob/b50b34f5611a1f182ba9b6eaf86c666bbd9f9eb0/sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala#L189-L214

SparkQA commented:
**[Test build #65865 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65865/consoleFull)** for PR 15231 at commit [`41611e5`](https://github.com/apache/spark/commit/41611e594c4e5f8e6aedb2846abfda74dac84a80).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65865/
Test FAILed.

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #65866 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65866/consoleFull)** for PR 15231 at commit [`41611e5`](https://github.com/apache/spark/commit/41611e594c4e5f8e6aedb2846abfda74dac84a80).

SparkQA commented:
**[Test build #65866 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65866/consoleFull)** for PR 15231 at commit [`41611e5`](https://github.com/apache/spark/commit/41611e594c4e5f8e6aedb2846abfda74dac84a80).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65866/
Test FAILed.

HyukjinKwon committed:
Fix documentation for CRAN check

SparkQA commented:
**[Test build #65868 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65868/consoleFull)** for PR 15231 at commit [`2d76e7c`](https://github.com/apache/spark/commit/2d76e7c54086dca703f4ffc180855c837bbc74e6).

SparkQA commented:
**[Test build #65868 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65868/consoleFull)** for PR 15231 at commit [`2d76e7c`](https://github.com/apache/spark/commit/2d76e7c54086dca703f4ffc180855c837bbc74e6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65868/
Test PASSed.

felixcheung commented:
Let's add some tests for this?

HyukjinKwon committed:
Add a test for calling DataFrameWriter.save() API without a path

SparkQA commented:
**[Test build #65880 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65880/consoleFull)** for PR 15231 at commit [`c2a64db`](https://github.com/apache/spark/commit/c2a64dba42d68561c23583ca8bbb79aa4edfb707).

HyukjinKwon commented:
@felixcheung Yeap, I just added. As we don't currently have an internal datasource allowing `DataFrameWriter.save()` without a path, I just added a test with CSV data source to make sure it throws an exception from Spark side (not from SparkR).

SparkQA commented:
**[Test build #65880 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65880/consoleFull)** for PR 15231 at commit [`c2a64db`](https://github.com/apache/spark/commit/c2a64dba42d68561c23583ca8bbb79aa4edfb707).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65880/
Test PASSed.

felixcheung commented:
Hmm, should we hold till 12601 is merged then? Seems like we shouldn't allow this unless internal datasources are supporting this more broadly.

Also, before the path parameter type is in the signature, ie.
```
write.df(df, c(1, 2))
```
Would error with some descriptive error, with this change it would get some JVM exception which seems to degrade the experience a bit.
Could you add checks for `path` and give more descriptive message?

HyukjinKwon commented:
@felixcheung , I usually don't like to answer by quote but let me do this just to clarify.

> Hmm, should we hold till 12601 is merged then? Seems like we shouldn't allow this unless internal datasources are supporting this more broadly.

As omitting `path` is what the datasource interface allows, maybe, it'd be just okay to test if it goes through JVM fine. Also, I worry if I can easily add a test for JDBC datasource within SparkR. If it can be easily done, I am also happy to hold this.

> Also, before the path parameter type is in the signature, ie.
> 
> ```
> write.df(df, c(1, 2))
> ```
> 
> Would error with some descriptive error, with this change it would get some JVM exception which seems to degrade the experience a bit.



Yeap, I could add some type checks

> Similarly for the path not specified case java.lang.IllegalArgumentException - we generally try to avoid JVM exception showing up if possible.

Also, yes. Maybe, we could avoid the direct JVM message after catching this and make it pretty within R just like PySpark does[1]. (although I am not sure if it sounds good in R).

> Could you add checks to path for these cases and give more descriptive messages?

Sure, I will try to address the points.

[1]https://github.com/apache/spark/blob/9a5071996b968148f6b9aba12e0d3fe888d9acd8/python/pyspark/sql/utils.py#L64-L80

HyukjinKwon commented:
Oh, BTW it seems `read.df` also seems not allowing this? I will try to test and fix here together if so.

HyukjinKwon committed:
Support omitting path for read.df and make the error messages better

SparkQA commented:
**[Test build #65890 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65890/consoleFull)** for PR 15231 at commit [`5c3d222`](https://github.com/apache/spark/commit/5c3d2228aeddf3a43888c170e00729ecbfe6e4bd).

SparkQA commented:
**[Test build #65890 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65890/consoleFull)** for PR 15231 at commit [`5c3d222`](https://github.com/apache/spark/commit/5c3d2228aeddf3a43888c170e00729ecbfe6e4bd).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65890/
Test FAILed.

HyukjinKwon committed:
Fix style

SparkQA commented:
**[Test build #65893 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65893/consoleFull)** for PR 15231 at commit [`1440195`](https://github.com/apache/spark/commit/1440195f49b346bde24843bf5c97360dc0488daf).

HyukjinKwon commented:
(I just updated the PR description too)

HyukjinKwon committed:
Remove unneeded dataframe

SparkQA commented:
**[Test build #65894 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65894/consoleFull)** for PR 15231 at commit [`07dca5d`](https://github.com/apache/spark/commit/07dca5d3d76b9e58dd1b2cdef8036c937a01f51b).

SparkQA commented:
**[Test build #65893 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65893/consoleFull)** for PR 15231 at commit [`1440195`](https://github.com/apache/spark/commit/1440195f49b346bde24843bf5c97360dc0488daf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65893/
Test PASSed.

SparkQA commented:
**[Test build #65894 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65894/consoleFull)** for PR 15231 at commit [`07dca5d`](https://github.com/apache/spark/commit/07dca5d3d76b9e58dd1b2cdef8036c937a01f51b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65894/
Test PASSed.

felixcheung commented on R/pkg/R/DataFrame.R:
```diff
@@ -2624,6 +2624,15 @@ setMethod("except",
           signature(df = "SparkDataFrame"),
           function(df, path = NULL, source = NULL, mode = "error", ...) {
+            if (!is.character(path) && !is.null(path)) {
+              stop("path should be charactor, null or omitted.")
```
"character"?

felixcheung replied:
same below

felixcheung replied:
minor point: is it more efficient to flip the checks, ie.
`if (!is.null(path) && !is.character(path))`, since path defaults to NULL?

felixcheung commented on R/pkg/R/DataFrame.R:
```diff
@@ -2624,6 +2624,15 @@ setMethod("except",
+              stop("path should be charactor, null or omitted.")
+            }
+            if (!is.character(source) && !is.null(source)) {
+              stop("source should be charactor, null or omitted. It is 'parquet' by default.")
```
strictly speaking, it's `spark.sql.sources.default` property, and when it is not set, then it is `parquet`

felixcheung commented on R/pkg/R/utils.R:
```diff
@@ -698,6 +698,21 @@ isSparkRShell <- function() {
 
+captureJVMException <- function(e) {
+  stacktrace <- as.character(e)
+  if (any(grep("java.lang.IllegalArgumentException: ", stacktrace))) {
```
are there cases where the IllegalArgument should be checked on the R side first to avoid the exception in the first place?

HyukjinKwon replied:
Thanks! @felixcheung I will address all other comments above. However, for this one, I was thinking hard but it seems not easy because we won't know if given data source is valid or not in R side first.

I might be able to do this only for internal data sources or known databricks datasources such as "redshift" or "xml" like.. creating a map for our internal data sources and then checking a path is given or not. However, I am not sure if it is a good idea to manage another list for datasources.

felixcheung replied:
I agree, I don't think we should couple the R code to the underlining data source implementations, and was not suggesting that :)

I guess I'm saying there are still many (other) cases where the parameters are unchecked and would be good if this check to convert JVM IllegalArgumentException is sufficient or more checks should be added to the R side.

HyukjinKwon replied:
Ah, I see. Yeap. This might be about best effort thing. I think I tried (if I am right) all combinations of parameters mssing/wrong in the APIs. One exceptional case for both APIs is, they throw an exception, `ClassCastException` when the extra options are wrongly typed, which I think we should check within R side and this will be handled in https://github.com/apache/spark/pull/15239
I might better open another PR for validating parameters across SparkR if you think it is okay.

felixcheung replied:
great, thanks - generally I'd prefer having parameter checks in R; though in this case I think we need balance the added code complicity and reduced usability (by checking more, it might fail where it didn't before).

so I'm not 100% sure we should add parameter checks all across the board.

HyukjinKwon replied:
Yeap, I do understand and will investigate it with keeping this in mind :)

felixcheung commented on R/pkg/R/utils.R:
```diff
@@ -698,6 +698,21 @@ isSparkRShell <- function() {
   grepl(".*shell\\.R$", Sys.getenv("R_PROFILE_USER"), perl = TRUE)
 }
 
+captureJVMException <- function(e) {
```
It'll be great to add some tests that would trigger tryCatch and this function?

felixcheung replied:
since there are 3 cases here and their handling seem nontrivial, could you add a test for each of the 3 exception main types?

HyukjinKwon committed:
Address some comments

HyukjinKwon committed:
Address comments more and add a unittest for captureJVMException

HyukjinKwon commented:
The commits I just pushed address the comments except for https://github.com/apache/spark/pull/15231#discussion_r80417904

SparkQA commented:
**[Test build #65905 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65905/consoleFull)** for PR 15231 at commit [`37480be`](https://github.com/apache/spark/commit/37480be0f76b7e4404c639e62137fd301989aaf3).

SparkQA commented:
**[Test build #65904 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65904/consoleFull)** for PR 15231 at commit [`11ae832`](https://github.com/apache/spark/commit/11ae83276b01977283051d32233932909ce113dd).

SparkQA commented:
**[Test build #65905 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65905/consoleFull)** for PR 15231 at commit [`37480be`](https://github.com/apache/spark/commit/37480be0f76b7e4404c639e62137fd301989aaf3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65905/
Test PASSed.

SparkQA commented:
**[Test build #65904 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65904/consoleFull)** for PR 15231 at commit [`11ae832`](https://github.com/apache/spark/commit/11ae83276b01977283051d32233932909ce113dd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65904/
Test PASSed.

felixcheung commented:
since #12601 is merged, could you check if this should be updated or tested?

HyukjinKwon committed:
Fetch upstream

SparkQA commented:
**[Test build #66169 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66169/consoleFull)** for PR 15231 at commit [`e6aeac3`](https://github.com/apache/spark/commit/e6aeac3f08ddfab7274bd1489bb7608778e3a51e).

HyukjinKwon commented:
I just make this updated and manually tested the jdbc datasource with h2 database as below 

with "text.json" below:

```json
{"name": "bcd"}
{"name": "abc"}
{"name": "Hyukjin"}
```

with the codes below:

```r
> df <- read.df(source="jdbc",
+               driver="org.h2.Driver",
+               url="jdbc:h2:mem:testdb0;user=testUser;password=testPass",
+               dbtable="PEOPLE")
> collect(df)
  NAME
1 fred
2 mary
> jsonDF <- read.json("test.json")
> write.df(jsonDF,
+         source = "jdbc",
+         mode = "overwrite",
+         driver = "org.h2.Driver",
+         url = "jdbc:h2:mem:testdb0;user=testUser;password=testPass",
+         dbtable = "PEOPLE")
> readback <- read.df(source = "jdbc",
+                     driver = "org.h2.Driver",
+                     url = "jdbc:h2:mem:testdb0;user=testUser;password=testPass",
+                     dbtable = "PEOPLE")
> collect(readback)
     name
1     bcd
2     abc
3 Hyukjin
```

HyukjinKwon commented:
@felixcheung BTW, it seems pretty difficult to add a test for JDBC. Maybe, I could do this later in another PR.

SparkQA commented:
**[Test build #66169 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66169/consoleFull)** for PR 15231 at commit [`e6aeac3`](https://github.com/apache/spark/commit/e6aeac3f08ddfab7274bd1489bb7608778e3a51e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66169/
Test PASSed.

felixcheung commented:
We talked about loading a jar that is running as h2 source, and then calling read.jdbc/write.jdbc from R.
Agreed that would be a separate project, but might be a good idea in the long run.

felixcheung commented:
I tested this, and this is the message I see:
```
16/10/02 05:43:17 ERROR RBackendHandler: getSQLDataType on org.apache.spark.sql.api.r.SQLUtils failed
Error in value[[3L]](cond) : Invalid type unknown
```

I think we should lose the first part, "in value[[3L]](cond)"? Perhaps we have the function name instead "getSQLDataType" instead?

Also I think it'd be important to differentiate where the message is coming from, so how about this add that to the stop calls, so how about something like

```
Error in getSQLDataType : illegal argument - Invalid type unknown
```

HyukjinKwon commented:
@felixcheung I see. Yes, I think that is a good idea. I just read the codes further. How about the one like the original message before such as ..

```
Error in invokeJava(isStatic = TRUE, className, methodName, ...): illegal argument - Invalid type unknown
```

for now and then dealing with printing the original name (`getSQLDataType`) in the future PR which I will submit (which we taked in https://github.com/apache/spark/pull/15231#discussion_r80594580)?

It seems one of the easy ways to reach the goal would be a wrapper for `callJMethod` and treating them in a single place with `captureJVMException`.

HyukjinKwon committed:
Improve error message

felixcheung commented:
I'd prefer `getSQLDataType` over `invokeJava` in the message - `invokeJava` is likely going to be in every exception and as such would not be useful for debugging.

Since this is regarding `captureJVMException` which is added here, I think it would be better to have the message change in this PR. It'd be important for that information in the message output because otherwise we are degrading the information we have when an user is reporting an error - it'd be very hard to troubleshoot `Error in value[[3L]](cond) : Invalid type unknown`. More R side argument checks as discussed in https://github.com/apache/spark/pull/15231#discussion_r80594580 would be ok to be logically in another PR/JIRA.

@HyukjinKwon How about we make the message change here and merge this PR ASAP? Am I understanding the context or level of the changes required for this?

HyukjinKwon commented:
@felixcheung ah, yes, maybe, I just wanted to avoid a change a lot but it seems making the change here is more sensible. Let me try to fix this here together. Thank you!

HyukjinKwon committed:
Address comments about exception messages

SparkQA commented:
**[Test build #66311 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66311/consoleFull)** for PR 15231 at commit [`bcd5060`](https://github.com/apache/spark/commit/bcd5060ba8aa7271f823cf60d4cc8cfb36850cf3).

HyukjinKwon commented:
@felixcheung I just updated this and now it'd print the exceptions as below:

```r
> read.df(source="json")
Error in loadDF : analysis error - Unable to infer schema for JSON at . It must be specified manually;

> read.df("arbitrary_path")
Error in loadDF : analysis error - Path does not exist: file: .../arbitrary_path;

> write.df(df, source="csv")
Error in save : illegal argument - 'path' is not specified
```

Could you please take another look?

HyukjinKwon committed:
Remove unused variable

SparkQA commented:
**[Test build #66312 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66312/consoleFull)** for PR 15231 at commit [`08bdc4d`](https://github.com/apache/spark/commit/08bdc4d95c26c152279982075130fe5edaf3e456).

HyukjinKwon committed:
Add missed changes more

SparkQA commented:
**[Test build #66313 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66313/consoleFull)** for PR 15231 at commit [`c8a433b`](https://github.com/apache/spark/commit/c8a433bf7b52b04451412adc5b7b67003a93de12).

SparkQA commented:
**[Test build #66311 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66311/consoleFull)** for PR 15231 at commit [`bcd5060`](https://github.com/apache/spark/commit/bcd5060ba8aa7271f823cf60d4cc8cfb36850cf3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66311/
Test PASSed.

SparkQA commented:
**[Test build #66312 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66312/consoleFull)** for PR 15231 at commit [`08bdc4d`](https://github.com/apache/spark/commit/08bdc4d95c26c152279982075130fe5edaf3e456).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66312/
Test PASSed.

SparkQA commented:
**[Test build #66313 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66313/consoleFull)** for PR 15231 at commit [`c8a433b`](https://github.com/apache/spark/commit/c8a433bf7b52b04451412adc5b7b67003a93de12).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66313/
Test PASSed.

felixcheung commented on R/pkg/R/utils.R:
```diff
@@ -698,18 +698,55 @@ isSparkRShell <- function() {
+    # "Error in invokeJava(...)". Here, it replaces the characters to
+    # `paste("Error in", method, ":")` in order to identify which function
+    # was called in JVM side.
+    stacktrace <- strsplit(rawmsg, "Error in .*?: ")[[1]]
```
very minor nit: you could probably replace the double pass with grep above and strsplit with just the result from strsplit

felixcheung commented on R/pkg/inst/tests/testthat/test_utils.R:
```diff
@@ -167,10 +167,13 @@ test_that("convertToJSaveMode", {
 test_that("captureJVMException", {
-  expect_error(tryCatch(callJStatic("org.apache.spark.sql.api.r.SQLUtils", "getSQLDataType",
+  method <- "getSQLDataType"
+  expect_error(tryCatch(callJStatic("org.apache.spark.sql.api.r.SQLUtils", method,
```
let's change this test to `handledCallJStatic` too in a follow up?

felixcheung commented:
I think also it would be great to add to examples one for read.df and one for write.df without the path parameter (like a jdbc one)

felixcheung commented:
I've merged this to master. We could take some example improvements along with SPARK-17665

HyukjinKwon commented:
Thanks again for your close review. Will keep in mind the comments.
